### PR TITLE
Restricted Pod E2E tests

### PR DESCRIPTION
**Pull request by tallclair**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

1. Add a new `MixinRestrictedPodSecurity` and corresponding `MustMixinRestrictedPodSecurity` utilities which set the required fields to pass the restricted pod security checks, and verify that the given pod does in fact meet the requirements. These are complementary to the existing `GetRestricted{Pod,Container}SecurityContext` functions, but in most cases require fewer changes.
2. Update the `test/e2e/common/node/pods.go` tests to set the restricted namespace label and use the new utilities.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/area test
/assign @liggitt @s-urbaniak

### CODE REVIEW

**Comment by tallclair** [comment]
Sorry, looks like I have a little more work to do on this.

**Comment by tallclair** [comment]
This should be ready for review, sorry about the noise.

File: test/e2e/framework/pod/utils.go [hunk 1]
```diff
@@ -140,14 +141,15 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
 var psaEvaluator, _ = psapolicy.NewEvaluator(psapolicy.DefaultChecks())
 
 // MustMixinRestrictedPodSecurity makes the given pod compliant with the restricted pod security level.
-// If doing so would overrwrite existing non-conformant conifguration, a test failure is triggered.
-func MustMixinRestrictedPodSecruity(pod *v1.Pod) {
+// If doing so would overrwrite existing non-conformant configuration, a test failure is triggered.
```
Comment (pohly) [inline test/e2e/framework/pod/utils.go:144]: nit: I like that you had split up this PR into two different commits, but then this typo fix belongs into the first commit.

Same with all other utils.go changes in the second commit.

File: test/e2e/framework/pod/utils.go [hunk 2]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
+			container.SecurityContext.Capabilities = &v1.Capabilities{}
+		}
+		if len(container.SecurityContext.Capabilities.Drop) == 0 {
+			container.SecurityContext.Capabilities.Drop = []v1.Capability{"ALL"}
+		}
+	}
+}
+
+// RestrictedPod returns a bare-bones pod that can run as restricted.
+func RestrictedPod(name string, imageID int) *v1.Pod {
```
Comment (pohly) [inline test/e2e/framework/pod/utils.go:203]: We probably need to wait for https://github.com/kubernetes/kubernetes/pull/109989 to be merged first and then change this, right?

**Review by pohly — COMMENTED** [review]
Looks good to me, but I don't know enough about PSA to do an in-depth review of the underlying logic.

File: test/e2e/framework/pod/utils.go [hunk 3]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
 		Capabilities:             &v1.Capabilities{Drop: []v1.Capability{"ALL"}},
 	}
 }
+
+var psaEvaluator, _ = psapolicy.NewEvaluator(psapolicy.DefaultChecks())
+
+// MustMixinRestrictedPodSecurity makes the given pod compliant with the restricted pod security level.
+// If doing so would overrwrite existing non-conformant configuration, a test failure is triggered.
+func MustMixinRestrictedPodSecruity(pod *v1.Pod) *v1.Pod {
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:145]:
```suggestion
func MustMixinRestrictedPodSecurity(pod *v1.Pod) *v1.Pod {
```

File: test/e2e/framework/pod/utils.go [hunk 4]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
+	err := MixinRestrictedPodSecruity(pod)
+	gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
+	return pod
+}
+
+// MixinRestrictedPodSecurity makes the given pod compliant with the restricted pod security level.
+// If doing so would overrwrite existing non-conformant configuration, an error is returned.
+func MixinRestrictedPodSecruity(pod *v1.Pod) error {
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:153]:
```suggestion
func MixinRestrictedPodSecurity(pod *v1.Pod) error {
```

File: test/e2e/framework/pod/utils.go [hunk 5]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
 	}
 }
+
+var psaEvaluator, _ = psapolicy.NewEvaluator(psapolicy.DefaultChecks())
+
+// MustMixinRestrictedPodSecurity makes the given pod compliant with the restricted pod security level.
+// If doing so would overrwrite existing non-conformant configuration, a test failure is triggered.
+func MustMixinRestrictedPodSecruity(pod *v1.Pod) *v1.Pod {
+	err := MixinRestrictedPodSecruity(pod)
+	gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:147]: if this trips, does it give context of where it was called from so the caller can figure out their issue?

File: test/e2e/framework/pod/utils.go [hunk 6]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
+	return pod
+}
+
+// MixinRestrictedPodSecurity makes the given pod compliant with the restricted pod security level.
+// If doing so would overrwrite existing non-conformant configuration, an error is returned.
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:152]:
```suggestion
// If doing so would overwrite existing non-conformant configuration, an error is returned.
```

File: test/e2e/framework/pod/utils.go [hunk 7]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
+	if container.SecurityContext == nil {
+		container.SecurityContext = GetRestrictedContainerSecurityContext()
+	} else {
+		if container.SecurityContext.AllowPrivilegeEscalation == nil {
+			container.SecurityContext.AllowPrivilegeEscalation = pointer.Bool(false)
+		}
+		if container.SecurityContext.Capabilities == nil {
+			container.SecurityContext.Capabilities = &v1.Capabilities{}
+		}
+		if len(container.SecurityContext.Capabilities.Drop) == 0 {
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:196]: can't we append ALL to a non-empty list if it doesn't already exist in the list?

File: test/e2e/framework/pod/utils.go [hunk 8]
```diff
@@ -130,3 +137,82 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
+		}
+	}
+}
+
+// RestrictedPod returns a bare-bones pod that can run as restricted.
+func RestrictedPod(name string, imageID int) *v1.Pod {
+	return MustMixinRestrictedPodSecruity(&v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: map[string]string{},
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:207]: no change required, but I'm curious why the non-nil labels?

File: test/e2e/framework/pod/utils.go [hunk 9]
```diff
@@ -119,6 +125,7 @@ func GetLinuxLabel() *v1.SELinuxOptions {
 func GetRestrictedPodSecurityContext() *v1.PodSecurityContext {
 	return &v1.PodSecurityContext{
 		RunAsNonRoot:   pointer.BoolPtr(true),
+		RunAsUser:      pointer.Int64(1000),
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:128]: was this required? I thought RunAsNonRoot was sufficient

are all the callers of this function ok with running their images as uid 1000?

File: test/e2e/common/node/pods.go [hunk 10]
```diff
@@ -184,7 +184,7 @@ func expectNoErrorWithRetries(fn func() error, maxRetries int, explain ...interf
 
 var _ = SIGDescribe("Pods", func() {
 	f := framework.NewDefaultFramework("pods")
-	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelBaseline
```
Comment (liggitt) [inline test/e2e/common/node/pods.go:187]: are all tests in this suite run in presubmit or are any skipped?

File: test/e2e/common/node/pods.go [hunk 11]
```diff
@@ -200,19 +200,7 @@ var _ = SIGDescribe("Pods", func() {
...
-			Spec: v1.PodSpec{
-				Containers: []v1.Container{
-					{
-						Name:  "test",
-						Image: imageutils.GetPauseImageName(),
-					},
-				},
-			},
-		})
+		testHostIP(podClient, e2epod.RestrictedPod(name, imageutils.Pause))
```
Comment (liggitt) [inline test/e2e/common/node/pods.go:203]: did this intentionally change from imageutils.GetPauseImageName()?

File: test/e2e/common/node/pods.go [hunk 12]
```diff
@@ -224,23 +212,8 @@ var _ = SIGDescribe("Pods", func() {
 		ginkgo.By("creating the pod")
 		name := "pod-submit-remove-" + string(uuid.NewUUID())
 		value := strconv.Itoa(time.Now().Nanosecond())
-		pod := &v1.Pod{
```
Comment (liggitt) [inline test/e2e/common/node/pods.go:227]: this would be easier to review if it simply wrapped with the MustMixinRestrictedPod call and left the rest of the pod alone

File: test/e2e/common/node/pods.go (hunk 11 above)
Comment (liggitt) [reply test/e2e/common/node/pods.go:203]: this would be easier to review if it simply wrapped with the MustMixinRestrictedPod call and left the rest of the pod alone

File: test/e2e/framework/pod/utils.go (hunk 8 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:207]: It seems pretty common for tests to add labels to the test pods. This way callers don't need to initialize the labels map before setting labels (I suppose they could also just set the labels map directly). I thought about adding an argument to pass in labels, but landed on this instead.

File: test/e2e/framework/pod/utils.go (hunk 1 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:144]: Ack. I'll clean this up before submitting.

**Comment by tallclair** [comment]
/hold
To clean up commits before submitting.

File: test/e2e/framework/pod/utils.go (hunk 2 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:203]: Yeah, this method was the motivation for that change. Synced & updated.

File: test/e2e/framework/pod/utils.go (hunk 3 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:145]: Ugh, this is what I get for overusing autocomplete. Thanks for catching it.

File: test/e2e/framework/pod/utils.go (hunk 7 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:196]: If an E2E test is explicitly adding individual capabilities to drop, I'd presume it's testing something specific related to that, so I think it's safer to ignore it in that case. Alternatively, it might be cargo-culting something in, in which case it's better to clean up the spec. In either case, I don't see much value to being too smart here.

File: test/e2e/framework/pod/utils.go (hunk 9 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:128]: It's not required as long as the image is configured to run as a non-root user, otherwise the container will fail at runtime.

It looks like `pause` (65535) and `nonroot` (1234) are the only 2 test images setting a user, so I'm inclined to keep this for now (neither would care about being run as user 1000). It would be good to update more test images to run as non-root though.

Inconsistency was just an oversight (fixed).

Comment (tallclair) [reply test/e2e/framework/pod/utils.go:128]: I also updated the unit test to catch inconsistencies like this.

File: test/e2e/common/node/pods.go (hunk 11 above)
Comment (tallclair) [reply test/e2e/common/node/pods.go:203]: Yes, this is equivalent under the hood (RestrictedPod passes the ID to `imageutils.GetE2EImage`, and `GetPauseImageName` just returns `imageutils.GetE2EImage(imageutils.Pause)`)

I decided that using the imageID was cleaner, since most of our tests use e2e images.

File: test/e2e/common/node/pods.go (hunk 12 above)
Comment (tallclair) [reply test/e2e/common/node/pods.go:227]: Ack. I've been thinking about adding a set of "PodBuilder" utilities for e2e tests. Maybe I should just defer the RestrictedPod stuff to that...

In a few of these cases, the image was changed from nginx. I spent a while trying to get nginx to run as non-root, but it really needs image changes to make it work, and these tests don't actually care what image is being run.

**Comment by tallclair** [comment]
(pushed some fixes, will get to the rest of the feedback this afternoon)

File: test/e2e/common/node/pods.go (hunk 10 above)
Comment (tallclair) [reply test/e2e/common/node/pods.go:187]: No, slow tests are skipped. I ran them against a `v1.23.5-gke.2400` (rapid channel) GKE cluster and they passed.

**Comment by tallclair** [comment]
Alright, all feedback addressed.

File: test/e2e/common/node/pods.go [hunk 13]
```diff
@@ -235,12 +235,12 @@ var _ = SIGDescribe("Pods", func() {
 			Spec: v1.PodSpec{
 				Containers: []v1.Container{
 					{
-						Name:  "nginx",
-						Image: imageutils.GetE2EImage(imageutils.Nginx),
+						Name:  "pause",
```
Comment (liggitt) [inline test/e2e/common/node/pods.go:238]: does changing from nginx to pause change the lifecycle characteristics of the pod (w.r.t. handling shutdown signals)? same question applies in multiple places

File: test/e2e/framework/pod/utils.go [hunk 14]
```diff
@@ -115,10 +120,13 @@ func GetLinuxLabel() *v1.SELinuxOptions {
 		Level: "s0:c0,c1"}
 }
 
+const nonRootUser = 1000
+
 // GetRestrictedPodSecurityContext returns a minimal restricted pod security context.
 func GetRestrictedPodSecurityContext() *v1.PodSecurityContext {
 	return &v1.PodSecurityContext{
 		RunAsNonRoot:   pointer.BoolPtr(true),
+		RunAsUser:      pointer.Int64(nonRootUser),
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:129]: I'm still unsure about this... this isn't really a minimal restricted pod security context, right? shouldn't `runAsNonRoot: true` be sufficient?

File: test/e2e/framework/pod/utils.go [hunk 15]
```diff
@@ -130,3 +138,67 @@ func GetRestrictedContainerSecurityContext() *v1.SecurityContext {
...
+func MixinRestrictedPodSecurity(pod *v1.Pod) error {
+	if pod.Spec.SecurityContext == nil {
+		pod.Spec.SecurityContext = GetRestrictedPodSecurityContext()
+	} else {
+		if pod.Spec.SecurityContext.RunAsNonRoot == nil {
+			pod.Spec.SecurityContext.RunAsNonRoot = pointer.BoolPtr(true)
+		}
+		if pod.Spec.SecurityContext.RunAsUser == nil {
+			pod.Spec.SecurityContext.RunAsUser = pointer.Int64Ptr(nonRootUser)
+		}
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:163]: I also think we should avoid this if runAsNonRoot is sufficient

File: test/e2e/common/node/pods.go (hunk 13 above)
Comment (tallclair) [reply test/e2e/common/node/pods.go:238]: Good question. Both the [pause container](https://github.com/kubernetes/kubernetes/blob/4bd396115def9afd7eac7f9071ed699d61c55b27/build/pause/linux/pause.c#L55-L62) and [nginx](http://nginx.org/en/docs/control.html) handle SIGINT and SIGTERM as a fast shutdown, which I think is what containerd uses. nginx additionally handles SIGQUIT, but I don't see any references to SIGQUIT in k/k, containerd, or cri-o.

Maybe @SergeyKanzhelev or @endocrimes have additional context.

File: test/e2e/framework/pod/utils.go (hunk 14 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go:129]: Part of my goal with these methods was that most e2e test authors can just apply `MixinRestrictedPodSecurity` and not think about the restricted profile at all. If we don't set `RunAsUser` here (and in the mixin) then for all cases other than the pause pod the test author will need to manually set a RunAsUser. They'll probably pick something random that gets cargo-culted around.

What's your argument against this?

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:129]: > If we don't set RunAsUser here (and in the mixin) then for all cases other than the pause pod the test author will need to manually set a RunAsUser

Would they *have* to set it, or do you mean because the container image would have to be built to run with a non-root uid, which is uncommon?

> What's your argument against this?

It would break images that expect to run with a particular non-zero uid at runtime, right?

Comment (tallclair) [reply test/e2e/framework/pod/utils.go:129]: I searched through our test image manifests, and the only ones that set a user are:
- pause: user doesn't matter
- nonroot: all the tests using this override the user anyway
- windows containers: don't support Restricted for now (added TODO)

I'd prefer that the tests which require a specific user are explicit about it, rather than having all the tests that don't have an opinion set something arbitrary.

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:129]: > rather than having all the tests that don't have an opinion set something arbitrary

but… isn't this setting something arbitrary (1000) on all tests that don't have an opinion?

Comment (tallclair) [reply test/e2e/framework/pod/utils.go:129]: The difference is that defaulting it in a helper function makes it more obvious that it's just part of the boilerplate to run the test, not an essential part of what's being tested. This is a general complaint that I have about our e2e tests - they often set a bunch of random fields on the resources, and it's not clear which are core to the test.

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:129]: Is runAsUser generally needed as boilerplate, though? I thought runAsNonRoot was supposed to be sufficient.

If an e2e wanted to exercise `runAsNonRoot` functionality with an image that did not run as root by default and didn't want to set a runAsUser, that would mean they couldn't use GetRestrictedPodSecurityContext or MustMixinRestrictedPodSecurity to do so, because those wouldn't *actually* set the minimal securityContext that is compatible with the restricted PSS.

Comment (tallclair) [reply test/e2e/framework/pod/utils.go:129]: 1. IMO, if a test is explicitly testing the runAsNonRoot functionality (with or without a runAsUser set), then the test should be explicitly writing & managing the SecurityContext, and not deferring that to a helper.
2. There should probably be 1-2 E2E tests explicitly checking runAsNonRoot behavior. I'd prefer to optimize these convenience methods for the common path, and force those couple of tests to be write the pod manually.

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:129]: Ok, fair enough. Make sure the docs don't claim to produce a minimal valid restricted securityContext and doc the uid chosen

File: test/e2e/common/node/pods.go (hunk 13 above)
Comment (endocrimes) [reply test/e2e/common/node/pods.go:238]: This test doesn't cover any corner cases of shutdown so this seems ok to me 👍

File: test/e2e/framework/pod/utils.go (hunk 14 above)
Comment (endocrimes) [reply test/e2e/framework/pod/utils.go:129]: This reasoning makes sense to me 👍  - A comment summarizing this would go a long way for future readers of the framework code here tho.

**Review by endocrimes — COMMENTED** [review]
/approve

looks good from a node pov, just needs the comment/doc updates

File: test/e2e/framework/pod/utils.go (hunk 14 above)
Comment (tallclair) [reply test/e2e/framework/pod/utils.go]: Updated comments.

**Comment by tallclair** [comment]
Added comments about the default non-root user. Also squashed commits & rebased.

File: test/e2e/framework/pod/utils.go [hunk 16]
```diff
@@ -115,10 +120,16 @@ func GetLinuxLabel() *v1.SELinuxOptions {
...
+// DefaultNonRootUser is the default user ID used for running restricted (non-root) containers.
+const DefaultNonRootUser = 1000
+
+// GetRestrictedPodSecurityContext returns a restricted pod security context.
+// This includes setting RunAsUser for convenience, to pass the RunAsNonRoot check.
+// Tests that require a specific user ID should override this.
 func GetRestrictedPodSecurityContext() *v1.PodSecurityContext {
 	return &v1.PodSecurityContext{
 		RunAsNonRoot:   pointer.BoolPtr(true),
+		RunAsUser:      pointer.Int64(DefaultNonRootUser),
```
Comment (liggitt) [inline test/e2e/framework/pod/utils.go:132]: what do windows nodes do if runAsUser is non-nil? do they ignore it or fail?

Comment (jsturtevant) [reply test/e2e/framework/pod/utils.go:132]: It is ignored by kubelet: https://github.com/kubernetes/kubernetes/blob/68fc207cd971336d0942cdb7c1d697284ad60eaf/pkg/kubelet/kuberuntime/security_context_windows.go#L46-L49

With PodOS feature and OS field is set (in beta in 1.24) the adminsion controller would reject it: https://github.com/kubernetes/kubernetes/blob/976a940f4a4e84fe814583848f97b9aafcdb083f/staging/src/k8s.io/api/core/v1/types.go#L3286-L3307

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:132]: ok, there's a TODO to avoid setting RunAsUser in MustMixinRestrictedPodSecurity for explicitly windows pods, I just wanted to make sure this wasn't going to fail in a windows kubelet for pods that didn't explicitly indicate their OS

Comment (jsturtevant) [reply test/e2e/framework/pod/utils.go:132]: So, this did break the tests because of https://github.com/kubernetes/kubernetes/issues/102849

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-capz-master-containerd-windows/1529769915945324544

```
May 26 10:47:49.396: INFO: At 2022-05-26 10:44:57 +0000 UTC - event for pod-update-d046cc17-42c7-493c-8c0c-fea0e9ac3164: {kubelet capz-conf-qwm24} FailedMount: (combined from similar events): MountVolume.SetUp failed for volume "kube-api-access-9l7hz" : chown c:\var\lib\kubelet\pods\0a6b5b55-fc8f-42b4-b285-19aedc311882\volumes\kubernetes.io~projected\kube-api-access-9l7hz\..2022_05_26_10_46_59.624505313\token: not supported by windows
```

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:132]: this isn't setting RunAsUserName, but does the windows kubelet have similar problems with RunAsUser?

if so, that seems like a good reason to drop RunAsUser here

Comment (jsturtevant) [reply test/e2e/framework/pod/utils.go:132]: Yes the title is slightly misleading but it states in the issue: `In addition if a Windows Pod is created with a Projected Volume and RunAsUser set, the Pod will be stuck at ContainerCreating:`

Comment (liggitt) [reply test/e2e/framework/pod/utils.go:132]: ok. @tallclair, that's a good reason not to set RunAsUser here for pods that could run on Windows. @jsturtevant, do you mind opening a PR to drop the RunAsUser field?

Comment (jsturtevant) [reply test/e2e/framework/pod/utils.go:132]: linking here for anyone following the thread https://github.com/kubernetes/kubernetes/pull/110235
